### ISSUE TITLE
[suggestion] Use HTTPlug

### ISSUE THREAD

**Issue by FinesseRus** [aws/aws-sdk-php#1381]
I suggest to use [HTTPlug](http://docs.php-http.org/en/latest/httplug/introduction.html) instead of Guzzle as a HTTP adapter. HTTPlug is a HTTP client interface. It would allow the AWS SDK library to be abstracted from a specific HTTP client implementation. Users would be able to choose [any HTTP client](https://packagist.org/providers/php-http/client-implementation) they want to use with the SDK.

**Comment by mtdowling** [comment]
> It would allow the AWS SDK library to be abstracted from a specific HTTP client implementation. Users would be able to choose any HTTP client they want to use with the SDK.

This is already supported. The AWS SDK has a pluggable HTTP client system. For example, it supports both Guzzle 5 and 6. You can configure the SDK to use HTTPlug today if you wanted.

https://github.com/aws/aws-sdk-php/tree/master/src/Handler

With regards to HTTPlug, Guzzle actually has the same goal of being transport agnostic, and it supports pluggable HTTP client implementations. For example, Guzzle could technically support HTTPlug as a handler. See http://docs.guzzlephp.org/en/stable/handlers-and-middleware.html

**Comment by FinesseRus** [comment]
The problem is that the AWS SDK makes me download Guzzle even if I don't use it. Many other libraries request different HTTP clients which leads to a bunch of HTTP clients in my `vendor` folder.

Choosing Guzzle as a HTTP client is a business decision. But the AWS SDK is a library, and by definition, a library aims to be plugged to a business project, it can't be used by its own. Developers of business projects make business decision, library authors must not make the decision for them. So libraries need be as abstract as possible.

It's true that Guzzle is the de-facto HTTP client for PHP and that it is extendable, but it became the first choice before the rise of HTTPlug, which is going to become a PSR standard. With PHPPlug developers will have the business decision about the implementation in their hands: a Zend project may use Diactoros, a micro-framework project may use Buzz, any other framework may use Guzzle and so on.

There are some benefits for the AWS SDK developers. HTTPlug is a very stable project because it is an interface, that's why it will not have as many major versions as Guzzle has, so you won't need to update the AWS SDK for a new HTTP client version so often. And you won't need to support several versions of HTTP clients simultaneously like you do with Guzzle 5 and 6.

**Comment by mtdowling** [comment]
Thanks for your feedback.

Guzzle is used in the SDK in three ways:

1. As a handler implementation; Guzzle just works on most systems as it supports curl and a fallback to PHP streams. This is `GuzzleHttp/Guzzle`.
2. As a PSR-7 implementation; Guzzle provides implementations of PSR-7 messages. This is `GuzzleHttp/Psr7`.
3. For higher level PSR-7 stream operations / decorators; Guzzle provides things like LimitStream that are super useful. This is also in This is `GuzzleHttp/Psr7`.

You're right that choosing Guzzle is both a technical and business decision. We chose to use Guzzle because it is widely used, has the features we need for the SDK, it is a project that I created and am familiar with (and I work on the SDK team), and because several AWS employees have commit access so we can fix critical issues quickly with little impediments.

> With PHPPlug developers will have the business decision about the implementation in their hands: a Zend project may use Diactoros, a micro-framework project may use Buzz, any other framework may use Guzzle and so on.

So this is where things are a bit nuanced -- Guzzle both has a client and implements PSR-7 messages. Perhaps you care more about the HTTP message implementation? One thing to keep in mind is that all of the public interfaces in the SDK typehint on PSR7 messages, not on Guzzle concretions. While we do create messages (i.e., Guzzle implementations of PSR-7 messages), we expose these in interfaces and typehints as PSR-7 messages, and downstream consumers shouldn't care about the implementation of the message. A library that expects a PSR-7 message doesn't care if it's implemented by Diactoros or Guzzle.

I understand your point that using Guzzle in the SDK for PSR-7 messages means we are pulling in a Guzzle dependency that might be yet another implementation of PSR-7 messages in your dependency closure. Can you expand on why this is an issue for you? Are you deploying to something like Lamdba and need a smaller application size?

> HTTPlug, which is going to become a PSR standard.

I don't think HTTPPlug is going to become a standard -- it might implement a standard. To be fair, Guzzle could also provide an implementation of a PSR client standard.

If there were a PSR HTTP client standard, then it's possible the SDK could be updated at some point to rely on that as the "handler" abstraction. Until then, I don't see a strong benefit in breaking our existing customers.

> There are some benefits for the AWS SDK developers. HTTPlug is a very stable project because it is an interface, that's why it will not have as many major versions as Guzzle has, so you won't need to update the AWS SDK for a new HTTP client version so often. And you won't need to support several versions of HTTP clients simultaneously like you do with Guzzle 5 and 6.

The client in the SDK can be swapped out today for any implementation you want, so the versioning of Guzzle isn't really a problem here. I should also point out that Guzzle hasn't had a breaking change in about 2.5 years. If I recall correctly, the PSR-7 package provided by Guzzle is still stable at 1.x.

In summary, I think we should circle back to this if/when a PSR comes out for HTTP clients. If that PSR meets the requirements of the AWS SDK for PHP, then we could look into how it can either be better supported by the SDK without a breaking change or supported in the next major version of the SDK.

**Comment by FinesseRus** [comment]
You use some Guzzle which are not provided by HTTPlug (as a HTTP client implementation), don't you? If you do, how can I switch Guzzle to other implementation?

> So this is where things are a bit nuanced -- Guzzle both has a client and implements PSR-7 messages. Perhaps you care more about the HTTP message implementation?

I am talking about HTTP client, not PSR-7 messages.

**Comment by kstich** [comment]
You would replace the detected Guzzle http handler that the SDK uses by default by setting the [`'handler'`](https://docs.aws.amazon.com/aws-sdk-php/v3/guide/guide/configuration.html#handler) or [`'http_handler'`](https://docs.aws.amazon.com/aws-sdk-php/v3/guide/guide/configuration.html#http-handler) option with an HTTPlug [handler implementation](https://docs.aws.amazon.com/aws-sdk-php/v3/guide/guide/handlers-and-middleware.html) that satisfies the specific interface. This would set the SDK to use HTTPlug through your handler.

As @mtdowling said above, if a PSR is finalized for HTTP clients, we can look in to supporting them.

**Comment by FinesseRus** [comment]
I got you point. Thank you for the explanation. Will be watching the destiny of the PSR HTTP client.
